### DISCUSSION TITLE
Why do I need an explicit call to sleep within wait?

### DISCUSSION THREAD

**Discussion by AshfordN** [fltk-rs/fltk-rs#978] [Q&A] [answered]
The [documentation](https://docs.rs/fltk/1.2.13/fltk/app/fn.wait.html) for `app::wait()` says:
> Calls to redraw within wait require an explicit sleep

But what does this mean exactly and why is it necessary?

**Comment by MoAlyousef** [comment, marked as answer]
FLTK's event loop normally sleeps until awoken by an event or a call to awake or a redraw.

Using: 
```rust
while app.wait() {
    win.redraw();
}
```
Would wake the event loop continuously thus leading to a busy loop.
So you would add an `app::sleep()` or `std::thread::sleep()` after `win.redraw();`. to prevent your app from using 100% cpu. 
If you don't have calls to redraw within the app.wait() loop, you don't need to sleep.

**Reply by AshfordN** [reply]
I've constructed a channel using `app::channel()` but I notice that some messages sent on the `Sender` don't make it out the other side, unless I place an explicit sleep between calls to `tx.send()`. This behaviour makes it seem as though the messages aren't being queued even though the [FLTK](https://www.fltk.org/doc-1.3/group__fl__multithread.html#gae9e8f440ce2ca05b047c620f75af13cb) docs say:

> Multiple calls to Fl::awake() will queue multiple pointers for the main thread to process, up to a system-defined (typically several thousand) depth.

Admittedly, though, the docs go on to say:

> The default message handler saves the last message which can be accessed using the Fl::thread_message() function.

and the description of the `thread_message()` function states that:

> The thread_message() method returns the last message that was sent from a child by the awake() method.

So, I'm a bit unclear as to how this works. Is queuing available, and is there a way to enable/access it?

**Reply by MoAlyousef** [reply]
FLTK doesn't give access to the message queue unfortunately. I've noticed also when sending messages from a handle methods, that the queue is bombarded with messages, and losing the last thread message makes it seems as though a message didn't make it out. 
For example sending a message on both Push and Released, the Released message might  cause the Push message to appear as if it had been dropped.
You can try a crossbeam-channel and see if the behavior is more favorable. Or you can save messages into a queue.

**Reply by AshfordN** [reply]
I followed your suggestion and came up with the following solution:

```rust
fn main() {
    // setup code

    let (tx, rx) = mpsc::channel();

    // event handling code

    while app::wait() {
        let queue = rx.try_iter();
        for msg in queue {
            // process message
        }
    }

    // clean up code
}
```

Perhaps you should consider documenting this little quirk, though. Unfortunately, my initial question did not properly explain the issue I was having, and others might struggle to find this solution.
